File: src/lib/tokens.ts
// Wrapper for the Rust-side keyring-backed token storage.
import { invoke } from '@tauri-apps/api/core';

export async function tokenSet(name: string, value: string): Promise<void> {
  await invoke('token_set', { name, value });
}

export async function tokenGet(name: string): Promise<string | null> {
  return (await invoke<string | null>('token_get', { name })) ?? null;
}

export async function tokenDelete(name: string): Promise<void> {
  await invoke('token_delete', { name });
}

File: src-tauri/src/tokens.rs
// Token storage backed by the native OS keychain (macOS Keychain / Windows
// Credential Manager) via the `keyring` crate. No master password prompts,
// no extra files on disk — just secure platform-native storage.

use keyring::Entry;

const SERVICE: &str = "com.centro.app";

fn entry(name: &str) -> Result<Entry, String> {
    Entry::new(SERVICE, name).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn token_set(name: String, value: String) -> Result<(), String> {
    entry(&name)?.set_password(&value).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn token_get(name: String) -> Result<Option<String>, String> {
    match entry(&name)?.get_password() {
        Ok(s) => Ok(Some(s)),
        Err(keyring::Error::NoEntry) => Ok(None),
        Err(e) => Err(e.to_string()),
    }
}

#[tauri::command]
pub fn token_delete(name: String) -> Result<(), String> {
    match entry(&name)?.delete_credential() {
        Ok(()) => Ok(()),
        Err(keyring::Error::NoEntry) => Ok(()),
        Err(e) => Err(e.to_string()),
    }
}
